api/handlers.go

```diff
@@ -5,42 +5,193 @@ import (
 			return
 		}
-		got, _, err := inst.Dispatch(r.Context(), "dataset.get", params)
+
+		switch returnCase {
+
+		case "zip":
+			outBytes, err := inst.Dataset().GetZip(r.Context(), params)
+			if err != nil {
+				util.RespondWithError(w, err)
+				return
+			}
+			writeFileResponse(w, outBytes, "dataset.zip", "zip")
+			return
+
+		case "csv":
+			outBytes, err := inst.Dataset().GetCSV(r.Context(), params)
+			if err != nil {
+				util.RespondWithError(w, err)
+				return
+			}
+			writeFileResponse(w, outBytes, "body.csv", "csv")
+			return
+
+		case "text/json":
+			res, err := inst.Dataset().Get(r.Context(), params)
+			if err != nil {
+				util.RespondWithError(w, err)
+			}
+			outbytes, err := json.Marshal(res.Value)
+			if err != nil {
+				util.RespondWithError(w, err)
+			}
+			componentName := params.Selector
+			if componentName == "" {
+				componentName = "dataset"
+			}
+			writeFileResponse(w, outbytes, fmt.Sprintf("%s.json", componentName), "json")
+			return
+
+		case "pretty":
+			res, err := inst.Dataset().Get(r.Context(), params)
+			if err != nil {
+				util.RespondWithError(w, err)
+			}
+			outbytes, err := json.MarshalIndent(res.Value, "", "  ")
+			if err != nil {
+				util.RespondWithError(w, err)
+			}
+			componentName := params.Selector
+			if componentName == "" {
+				componentName = "dataset"
+			}
+			writeFileResponse(w, outbytes, fmt.Sprintf("%s.json", componentName), "json")
+			return
+		}
+
+		res, err := inst.Dataset().Get(r.Context(), params)
 		if err != nil {
 			util.RespondWithError(w, err)
 			return
 		}
-		res, ok := got.(*lib.GetResult)
-		if !ok {
-			util.RespondWithDispatchTypeError(w, got)
+
+		if lib.IsSelectorScriptFile(params.Selector) {
+			util.WriteResponse(w, res.Bytes)
 			return
 		}
-		replyWithGetResponse(w, r, params, res)
+
+		util.WriteResponse(w, res.Value)
+	}
+}
+
+// possible return cases:
+//  - "zip"
+//  - "csv"
+//  - "pretty"
+//  - "text/json"
+//  - ""
+func getCaseAndParamsFromRequest(r *http.Request) (string, *lib.GetParams, error) {
+	p := &lib.GetParams{}
+	log.Debugf("ref:%s", r.FormValue("ref"))
+	p.Ref = r.FormValue("ref")
+
+	ref, err := dsref.Parse(p.Ref)
+	if err != nil {
+		return "", nil, err
+	}
+
+	if ref.Username == "me" {
+		return "", nil, fmt.Errorf("username \"me\" not allowed")
+	}
+
+	format := r.FormValue("format")
+	p.Selector = r.FormValue("selector")
+
+	p.All = util.ReqParamBool(r, "all", false)
+
+	// TODO(arqu): we default to true but should implement a guard and/or respect the page params
+	listParams := lib.ListParamsFromRequest(r)
+	offset := listParams.Offset
+	limit := listParams.Limit
+	if offset == 0 && limit == -1 {
+		p.All = true
+	}
+
+	switch {
+	case format == "csv":
+		fallthrough
+	case arrayContains(r.Header["Accept"], "text/csv"):
+		if !(format == "csv" || format == "") {
+			return "", nil, fmt.Errorf("format %q conflicts with header %q", format, "Accept: text/csv")
+		}
+		fallthrough
+	case p.Selector == "body.csv":
```

Comment:
In "api/api.go" line 163 we have `handleRefRoute` that ensures we handle: "/ds/get/{username}/{name}/{selector}"
We have `refStringMiddleware` that takes a ref string from the muxVars and adds a query param "ref" to the request. We have `setMuxVarsToQueryParams` that takes all the mux vars and sets them as query params on the request.

So if we ping url "/ds/get/my_username/my_dataset/body.csv", "body.csv" get's seen as the `selector`.

I agree this is ambiguous & muddled and I don't love it. Part of the issues is I initially had `getOutfileTypeAndParamsFromRequest` as two functions, one that parsed the request and set the `GetParams` and one that set the `outfileType`. However, as you can see from the section you pointed out, sometimes what ends up in the `Selector` field, actually represents something beyond just being a selector.

One thought I had was to use a different term than "selector" in `handleRefRoute` & `setMuxVarsToQueryParams` so it is clear that what we get from this value may not be a traditional "selector". I just didn't know what to call it.

Regardless of if we change the term in the route, I shouldn't be setting `p.Selector` until the value has been de-coupled from special cases:
```
func getOutfileTypeAndParamsFromRequest(r *http.Request) (outfileType, *lib.GetParams, error) {
  // parse request into params
  maybeSelector := r.FormValue("selector")
  // go through different outfileType options & adjust `maybeSelector` if it's one of our special cases
  p.Selector = maybeSelector
  return outfileType, p, nil
}
```

Reply:
Could we use this?

```
"/ds/get/{username}/{name}/{componentFile}"
```

Let's say we go with that. If there any valid value for `componentFile` that we need to support aside from "body.csv"? I know users specifically asked for that one, are there others? If not, I'd prefer just checking this `componentFile` field against the one allowed value, and disallowing anything else. That is, seeing `componentFile == "body.csv"` would set `selector="body"`, and `format="csv"`, etc.

If a user is requesting some specific component or field, they should still be able to do:

```
"/ds/get/{username}/{name}?selector={selector}"
```

such as:

```
"/ds/get/dustmp/my_ds?selector=structure.field"
```

Does that work?

Reply:
Yea, the issue is that one of the sugar cases we definitely want to support IS having the selector as part of the url:

```
"/ds/get/ramfox/my_dataset/meta" // returns just the meta component
```

Which is why `selector` was used in the first place, but... again... isn't always true.

Reply:
Ok, wasn't aware that also should be allowed. I thought part of the api break was removing this type of ref in the url, or at least requiring callers to use something like `"/ds/get/ramfox/my_dataset?selector=meta"`.

If `componentFile` is too specific, I'd go with something general like `path`.

Reply:
(Post-edit follow-up): I'm ok with handling it in whatever way is simplest. I guess my main concerns are:

1) Not adding features we don't need to, since there's an maintenance cost. I think /dataset.json are /meta.json etc are unnecessary, unless I missed some explicit need for them. Pretty sure they're just holdovers from the original api.
2) Keeping the term "selector" accurate in how it is used, and not allowing values like "body.csv", which are filenames not selectors.

With those in mind, my vote would be using either `path` or `componentFile`, whichever you think makes more sense.

Reply:
After taking a step back, I think the best most logical way to handle this is for us to keep the "selector" field pure, keep it as "selector", and handle "/ds/get/{username}/{name}/body.csv" separately.

All other "special cases" should be removed from this pr. As you mentioned above, there is a maintenance cost for every additional case & it is easier to add a case when a user requests it then it is to take a case away/maintain an old case. After reviewing these the special cases besides "body.csv", they aren't in our 0.10.0 endpoint list, I just got excited while coding and didn't keep in mind the maintenance cost.

This also dramatically reduces the complexity of the `GetHandler`, since there are way fewer "gotchyas" and expected return types.

Reply:
Sounds good! Reducing complexity is excellent!